tracer/src/Datadog.Trace/AppSec/EventTrackingSdkV2.cs

```diff
@@ -0,0 +1,242 @@
+    public static void TrackUserLoginSuccess(string userLogin, string? userId = null, IDictionary<string, string>? metadata = null)
+    {
+        TelemetryFactory.Metrics.Record(PublicApiUsage.EventTrackingSdkV2_TrackUserLoginSuccess_UserId);
+        TrackUserLoginSuccess(userLogin, userId is not null ? new UserDetails(userId) : null, metadata, Tracer.Instance);
+    }
+
+    /// <summary>
+    /// Sets the details of a successful logon on the local root span
+    /// </summary>
+    /// <param name="userLogin">The userId associated with the login success</param>
+    /// <param name="userDetails">The userLogin associated with the login success</param>
+    /// <param name="metadata">Metadata associated with the login success</param>
+    [PublicApi]
+    public static void TrackUserLoginSuccess(string userLogin, UserDetails userDetails, IDictionary<string, string>? metadata = null)
+    {
+        TelemetryFactory.Metrics.Record(PublicApiUsage.EventTrackingSdkV2_TrackUserLoginSuccess_UserDetails);
+        TrackUserLoginSuccess(userLogin, userDetails, metadata, Tracer.Instance);
+    }
+
+    /// <summary>
+    /// Sets the details of a logon failure on the local root span
+    /// </summary>
+    /// <param name="userLogin">The userId associated with the login failure</param>
+    /// <param name="exists">If the userId associated with the login failure exists</param>
+    /// <param name="userId">The optional userId associated with the login success</param>
+    /// <param name="metadata">Metadata associated with the login failure</param>
+    [PublicApi]
+    public static void TrackUserLoginFailure(string userLogin, bool exists, string? userId = null, IDictionary<string, string>? metadata = null)
+    {
+        TelemetryFactory.Metrics.Record(PublicApiUsage.EventTrackingSdkV2_TrackUserLoginFailure_UserId);
+        TrackUserLoginFailure(userLogin, exists, userId is not null ? new UserDetails(userId) : null, metadata, Tracer.Instance);
+    }
+
+    /// <summary>
+    /// Sets the details of a logon failure on the local root span
+    /// </summary>
+    /// <param name="userLogin">The user login associated with the login failure</param>
+    /// <param name="exists">If the userId associated with the login failure exists</param>
+    /// <param name="userDetails">The details of the user associated with the login failure</param>
+    /// <param name="metadata">Metadata associated with the login failure</param>
+    [PublicApi]
+    public static void TrackUserLoginFailure(string userLogin, bool exists, UserDetails userDetails, IDictionary<string, string>? metadata = null)
+    {
+        TelemetryFactory.Metrics.Record(PublicApiUsage.EventTrackingSdkV2_TrackUserLoginFailure_UserDetails);
+        TrackUserLoginFailure(userLogin, exists, userDetails, metadata, Tracer.Instance);
+    }
+
+    internal static void TrackUserLoginSuccess(string userLogin, IUserDetails? userDetails, IDictionary<string, string>? metadata, Tracer tracer)
```

Comment:
nit: I think you can avoid the boxing from the `IUserDetails?` by adding the generic constraint that you added in the automatic instrumentation classes: `where TUserDetails : IUserDetails`

This should avoid having to box this object when calling methods on it.

Reply:
thanks! fixed in bf0b85b385f29508dc4d0406e0770d77d13b1ab0